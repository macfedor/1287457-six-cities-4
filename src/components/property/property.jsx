import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {formatRating} from "../../utils/common.js";
import {PlaceType} from "../../consts.js";
import ReviewsList from "../reviews-list/reviews-list.jsx";
import NearbyPlaces from "../nearby-places/nearby-places.jsx";
import Map from "../map/map.jsx";
import offers from "../../mocks/offers.js";
import Header from "../header/header.jsx";
import {getAuthorizationStatus} from "../../reducer/user/selectors.js";
import {connect} from "react-redux";
import {Operation as DataOperation} from "../../reducer/data/data.js";
import {getReviewsList, getActiveOffer, getNearbyPlacesList} from "../../reducer/data/selectors.js";

class Property extends PureComponent {

  componentDidMount() {
    const {property, getReviews, getNearbyPlaces} = this.props;

    if (property) {
      getReviews(property.id);
      getNearbyPlaces(property.id);
    }
  }

  componentDidUpdate() {
    const {property, getReviews, getNearbyPlaces, reviews, nearbyPlaces} = this.props;

    if (property) {
      if (reviews === null) {
        getReviews(property.id);
      } 
      if (nearbyPlaces === null) {
        getNearbyPlaces(property.id);
      }
    }
  }

  render() {
    const mapPrefix = `property`;
    const {property, authorizationStatus, reviews, onFavoriteToggle, getOfferById, nearbyPlaces, onCardHover} = this.props;
    if (!property && this.props.routerProps.match.params.id) {
      getOfferById(this.props.routerProps.match.params.id);
    }
    if (!property) {
      return null;
    }

    return (
      <div className="page">
        <Header />

        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">
              <div className="property__gallery">
                {property.images.map((image, i) => (
                  <div key={`${i}-${image}`} className="property__image-wrapper">
                    <img className="property__image" src={image} alt="Photo studio" />
                  </div>
                ))}
              </div>
            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                {property.isPremium ?
                  <div className="property__mark">
                    <span>Premium</span>
                  </div>
                  : ``}
                <div className="property__name-wrapper">
                  <h1 className="property__name">
                    {property.name}
                  </h1>
                  <button onClick={() => {
                    onFavoriteToggle(property.id, Number(!property.isFavorite));
                  }} className={`${property.isFavorite ? `property__bookmark-button--active` : ``} property__bookmark-button button`} type="button">
                    <svg className="property__bookmark-icon" width="31" height="33">
                      <use xlinkHref="#icon-bookmark"></use>
                    </svg>
                    <span className="visually-hidden">To bookmarks</span>
                  </button>
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: formatRating(property.rating)}} />
                    <span className="visually-hidden">Rating</span>
                  </div>
                  <span className="property__rating-value rating__value">{property.rating}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">
                    {property.type}
                  </li>
                  <li className="property__feature property__feature--bedrooms">
                    {property.bedrooms} Bedrooms
                  </li>
                  <li className="property__feature property__feature--adults">
                    Max {property.guests} adults
                  </li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">&euro;{property.price}</b>
                  <span className="property__price-text">&nbsp;night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">What&apos;s inside</h2>
                  <ul className="property__inside-list">
                    {property.insideItems.map((item, i) => (
                      <li key={`${i}-${item}`} className="property__inside-item">
                        {item}
                      </li>
                    ))}
                  </ul>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div className={property.host.isPro ? `property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper` : `property__avatar-wrapper user__avatar-wrapper`}>
                      <img className="property__avatar user__avatar" src={`/${property.host.avatar}`} width="74" height="74" alt="Host avatar"/>
                    </div>
                    <span className="property__user-name">
                      {property.host.name}
                    </span>
                  </div>
                  <div className="property__description">
                    <p className="property__text">{property.description}</p>
                  </div>
                </div>
                {reviews !== null ?
                  <ReviewsList
                    reviews={reviews}
                    authorizationStatus={authorizationStatus}
                    propertyId={property.id}
                  />
                : ``}
              </div>
            </div>
            <Map
              places={nearbyPlaces}
              prefix={mapPrefix}
              activePlace={property}
            />
          </section>
          {nearbyPlaces !== null ?
            <div className="container">
              <NearbyPlaces
                places={nearbyPlaces}
                onCardHover={onCardHover}
              />
            </div>
            : ``}
        </main>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  property: getActiveOffer(state),
  reviews: getReviewsList(state),
  nearbyPlaces: getNearbyPlacesList(state),
});

const mapDispatchToProps = (dispatch) => ({
  getReviews(hotelId) {
    dispatch(DataOperation.loadReviews(hotelId));
  },
  getNearbyPlaces(hotelId) {
    dispatch(DataOperation.loadNearbyPlaces(hotelId));
  },
  onFavoriteToggle(id, status) {
    dispatch(DataOperation.toggleFavorite(id, status));
  },
  getOfferById(offerId) {
    dispatch(DataOperation.getOfferById(offerId));
  }
});

Property.propTypes = {
  property: PropTypes.shape({
    id: PropTypes.number.isRequired,
    image: PropTypes.string.isRequired,
    isPremium: PropTypes.bool.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    price: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.oneOf([PlaceType.APARTMENT, PlaceType.ROOM, PlaceType.HOUSE, PlaceType.HOTEL]).isRequired,
    rating: PropTypes.number.isRequired,
    images: PropTypes.arrayOf(PropTypes.string).isRequired,
    insideItems: PropTypes.arrayOf(PropTypes.string).isRequired,
    bedrooms: PropTypes.number.isRequired,
    guests: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    host: PropTypes.shape({
      name: PropTypes.string.isRequired,
      avatar: PropTypes.string.isRequired,
      isPro: PropTypes.bool.isRequired,
      id: PropTypes.number.isRequired,
    }).isRequired,
    location: PropTypes.shape({
      coordinates: PropTypes.arrayOf(PropTypes.number).isRequired,
      zoom: PropTypes.number.isRequired,
    }).isRequired,
    city: PropTypes.shape({
      name: PropTypes.string.isRequired,
      coordinates: PropTypes.arrayOf(PropTypes.number).isRequired,
      zoom: PropTypes.number.isRequired,
    }).isRequired,
  }),
  getReviews: PropTypes.func.isRequired,
  getOfferById: PropTypes.func.isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  reviews: PropTypes.arrayOf(PropTypes.shape({
    avatar: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    id: PropTypes.number.isRequired,
    date: PropTypes.string.isRequired,
    comment: PropTypes.string.isRequired
  })),
  onFavoriteToggle: PropTypes.func.isRequired,
};

export {Property};
export default connect(mapStateToProps, mapDispatchToProps)(Property);
