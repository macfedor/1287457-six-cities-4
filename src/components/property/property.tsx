import * as React from "react";
import {RouteProps} from "react-router-dom";
import {formatRating} from "../../utils/common";
import ReviewsList from "../reviews-list/reviews-list";
import NearbyPlaces from "../nearby-places/nearby-places";
import Map from "../map/map";
import Header from "../header/header";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import {connect} from "react-redux";
import {Operation as DataOperation} from "../../reducer/data/data";
import {getReviewsList, getActiveOffer, getNearbyPlacesList} from "../../reducer/data/selectors";
import {Offer, ReviewInterface} from "../../types";

type Props = RouteProps & {
  property: Offer;
  getReviews: (id: number) => void;
  onCardHover: () => void;
  getNearbyPlaces: (id: number) => void;
  getOfferById: (id: number) => void;
  authorizationStatus: string;
  reviews: ReviewInterface[];
  onFavoriteToggle: (id: number, status: number) => void;
  nearbyPlaces: Offer[];
}

class Property extends React.PureComponent<Props, Record<string, unknown>> {

  componentDidMount() {
    const {property, getReviews, getNearbyPlaces} = this.props;

    if (property) {
      getReviews(property.id);
      getNearbyPlaces(property.id);
    }
  }

  componentDidUpdate(prevProps) {
    const {property, getReviews, getNearbyPlaces, reviews, nearbyPlaces} = this.props;

    if (property) {
      if (reviews === null || prevProps.property && property.id !== prevProps.property.id) {
        getReviews(property.id);
      }
      if (nearbyPlaces === null || prevProps.property && property.id !== prevProps.property.id) {
        getNearbyPlaces(property.id);
      }
    }

    if (prevProps.property && property.id !== prevProps.property.id) {
      window.scrollTo(0, 0);
    }
  }

  render() {
    const mapPrefix = `property`;
    const {property, authorizationStatus, reviews, onFavoriteToggle, getOfferById, nearbyPlaces} = this.props;
    if (!property && this.props.routerProps.match.params.id) {
      getOfferById(this.props.routerProps.match.params.id);
    }
    if (!property) {
      return null;
    }

    return (
      <div className="page">
        <Header />

        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">
              <div className="property__gallery">
                {property.images.map((image, i) => (
                  <div key={`${i}-${image}`} className="property__image-wrapper">
                    <img className="property__image" src={image} alt="Photo studio" />
                  </div>
                ))}
              </div>
            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                {property.isPremium ?
                  <div className="property__mark">
                    <span>Premium</span>
                  </div>
                  : ``}
                <div className="property__name-wrapper">
                  <h1 className="property__name">
                    {property.name}
                  </h1>
                  <button onClick={() => {
                    onFavoriteToggle(property.id, Number(!property.isFavorite));
                  }} className={`${property.isFavorite ? `property__bookmark-button--active` : ``} property__bookmark-button button`} type="button">
                    <svg className="property__bookmark-icon" width="31" height="33">
                      <use xlinkHref="#icon-bookmark"></use>
                    </svg>
                    <span className="visually-hidden">To bookmarks</span>
                  </button>
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: formatRating(property.rating)}} />
                    <span className="visually-hidden">Rating</span>
                  </div>
                  <span className="property__rating-value rating__value">{property.rating}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">
                    {property.type}
                  </li>
                  <li className="property__feature property__feature--bedrooms">
                    {property.bedrooms} Bedrooms
                  </li>
                  <li className="property__feature property__feature--adults">
                    Max {property.guests} adults
                  </li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">&euro;{property.price}</b>
                  <span className="property__price-text">&nbsp;night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">What&apos;s inside</h2>
                  <ul className="property__inside-list">
                    {property.insideItems.map((item, i) => (
                      <li key={`${i}-${item}`} className="property__inside-item">
                        {item}
                      </li>
                    ))}
                  </ul>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div className={property.host.isPro ? `property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper` : `property__avatar-wrapper user__avatar-wrapper`}>
                      <img className="property__avatar user__avatar" src={`/${property.host.avatar}`} width="74" height="74" alt="Host avatar"/>
                    </div>
                    <span className="property__user-name">
                      {property.host.name}
                    </span>
                  </div>
                  <div className="property__description">
                    <p className="property__text">{property.description}</p>
                  </div>
                </div>
                {reviews !== null ?
                  <ReviewsList
                    reviews={reviews}
                    authorizationStatus={authorizationStatus}
                    propertyId={property.id}
                  />
                  : ``}
              </div>
            </div>
            <Map
              places={nearbyPlaces}
              prefix={mapPrefix}
              activePlace={property}
            />
          </section>
          {nearbyPlaces !== null ?
            <div className="container">
              <NearbyPlaces
                places={nearbyPlaces}
              />
            </div>
            : ``}
        </main>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  property: getActiveOffer(state),
  reviews: getReviewsList(state),
  nearbyPlaces: getNearbyPlacesList(state),
});

const mapDispatchToProps = (dispatch) => ({
  getReviews(hotelId) {
    dispatch(DataOperation.loadReviews(hotelId));
  },
  getNearbyPlaces(hotelId) {
    dispatch(DataOperation.loadNearbyPlaces(hotelId));
  },
  onFavoriteToggle(id, status) {
    dispatch(DataOperation.toggleFavorite(id, status));
  },
  getOfferById(offerId) {
    dispatch(DataOperation.getOfferById(offerId));
  }
});

export {Property};
export default connect(mapStateToProps, mapDispatchToProps)(Property);
