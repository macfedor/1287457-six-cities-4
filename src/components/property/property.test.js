import React from "react";
import renderer from "react-test-renderer";
import {Property} from "./property.jsx";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {Router} from "react-router-dom";
import history from "../../history.js";

const mockStore = configureStore([]);

const div = document.createElement(`div`);
div.id = `map`;
document.body.appendChild(div);

const testData = {
  id: Math.random(),
  image: `img/apartment-01.jpg`,
  isPremium: false,
  price: 1000,
  name: `Secont`,
  type: `room`,
  rating: 5,
  images: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`],
  insideItems: [`Wi-Fi`, `Washing machine`, `Towels`],
  bedrooms: 3,
  guests: 4,
  description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
  host: {
    name: `Angelina`,
    avatar: `img/avatar-angelina.jpg`,
    isPro: true,
    id: 2,
  },
  location: {
    coordinates: [52.3909553943508, 4.929309666406198],
    zoom: 13,
  },
  city: {
    coordinates: [52.3909553943508, 4.929309666406198],
    zoom: 13,
    name: `Paris`,
  },
};

const mockReviews = [
  {
    id: 1,
    avatar: `img/avatar-max.jpg`,
    name: `Max`,
    rating: 2.4,
    date: String(new Date(`2020-03-21`)),
    comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`
  },
  {
    id: 2,
    avatar: `img/avatar-angelina.jpg`,
    name: `Angelina`,
    rating: 5,
    date: String(new Date(`2020-04-23`)),
    comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`
  }
];

const store = mockStore({
  DATA: {
    step: `property`,
    activeCity: null,
    activeOffer: testData,
    hoveredOffer: null,
    cities: [],
    places: [],
    activeSortType: `popular`,
  },
  USER: {
    AuthorizationStatus: `NO_AUTH`,
    userEmail: ``,
  }
});

it(`Should Property render correctly`, () => {
  const tree = renderer
    .create(
        <Router history={history} >
          <Provider store={store}><Property
            property={testData}
            reviews={mockReviews}
            onTitleClick={() => {}}
            getReviews={() => {}}
          /></Provider>
        </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
