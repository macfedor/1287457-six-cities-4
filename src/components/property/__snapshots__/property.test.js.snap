// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Property render correctly 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__user-name user__name"
                >
                  
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/room.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/apartment-01.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/apartment-02.jpg"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Secont
            </h1>
            <button
              className=" property__bookmark-button button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="property__bookmark-icon"
                height="33"
                width="31"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              5
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              room
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              3
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              4
               adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              1000
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Wi-Fi
              </li>
              <li
                className="property__inside-item"
              >
                Washing machine
              </li>
              <li
                className="property__inside-item"
              >
                Towels
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="img/avatar-angelina.jpg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <h2
              className="reviews__title"
            >
              Reviews · 
              <span
                className="reviews__amount"
              >
                2
              </span>
            </h2>
            <ul
              className="reviews__list"
            >
              <li
                className="reviews__item"
              >
                <div
                  className="reviews__user user"
                >
                  <div
                    className="reviews__avatar-wrapper user__avatar-wrapper"
                  >
                    <img
                      alt="Reviews avatar"
                      className="reviews__avatar user__avatar"
                      height="54"
                      src="img/avatar-max.jpg"
                      width="54"
                    />
                  </div>
                  <span
                    className="reviews__user-name"
                  >
                    Max
                  </span>
                </div>
                <div
                  className="reviews__info"
                >
                  <div
                    className="reviews__rating rating"
                  >
                    <div
                      className="reviews__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "40%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <p
                    className="reviews__text"
                  >
                    A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.
                  </p>
                  <time
                    className="reviews__time"
                    dateTime="2020-02-21"
                  >
                    March 2020
                  </time>
                </div>
              </li>
              <li
                className="reviews__item"
              >
                <div
                  className="reviews__user user"
                >
                  <div
                    className="reviews__avatar-wrapper user__avatar-wrapper"
                  >
                    <img
                      alt="Reviews avatar"
                      className="reviews__avatar user__avatar"
                      height="54"
                      src="img/avatar-angelina.jpg"
                      width="54"
                    />
                  </div>
                  <span
                    className="reviews__user-name"
                  >
                    Angelina
                  </span>
                </div>
                <div
                  className="reviews__info"
                >
                  <div
                    className="reviews__rating rating"
                  >
                    <div
                      className="reviews__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "100%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <p
                    className="reviews__text"
                  >
                    A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.
                  </p>
                  <time
                    className="reviews__time"
                    dateTime="2020-03-23"
                  >
                    April 2020
                  </time>
                </div>
              </li>
            </ul>
            
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      >
        <div
          id="map"
          style={
            Object {
              "height": "100%",
            }
          }
        />
      </section>
    </section>
    
  </main>
</div>
`;
